src/http.cc

```diff
@@ -1933,6 +2071,49 @@ HttpStateData::httpBuildRequestHeader(HttpRequest * request,
     strConnection.clean();
 }
 
+void
+HttpStateData::makeUpgradeHeaders(HttpHeader &hdr_out)
+{
+    // We need virgin request received by squid. This is stored to
+    // AccessLogEntry::request member.
+    const HttpHeader &hdr_in = fwd->al->request ? fwd->al->request->header :request->header;
+
+    if (!hdr_in.has(Http::HdrType::UPGRADE))
+        return;
+
+    const String upgradeIn = hdr_in.getList(Http::HdrType::UPGRADE);
+    String upgradeOut;
+    const char *pos = nullptr;
+    const char *item;
+    int ilen = 0;
+    while (strListGetItem(&upgradeIn, ',', &item, &ilen, &pos)) {
+        auto it = std::find_if(Config.http_upgrade_protocols->begin(), Config.http_upgrade_protocols->end(), MatchProtocol(item, ilen));
```

Comment:
I am not sure it will be less complex. Actually required exactly the same parts used by MatchProtocol. And the MatchProtocol actually is about 20 lines of code with brackets and class definition. Why it is complex?
Away of that I think we have overengineer this part of code.